### PR TITLE
add provide/inject on functional context

### PR DESCRIPTION
**Pull request by Kingwl**
#5194

### CODE REVIEW

File: src/core/vdom/create-component.js [hunk 1]
```diff
@@ -183,11 +184,13 @@ function createFunctionalComponent (
   // gets a unique context - this is necessary for correct named slot check
   const _context = Object.create(context)
   const h = (a, b, c, d) => createElement(_context, a, b, c, d, true)
+  const inject = resolveInject(Ctor.options.inject, _context)
   const vnode = Ctor.options.render.call(null, h, {
     props,
     data,
     parent: context,
     children,
+    inject,
```
Comment (yyx990803) [inline src/core/vdom/create-component.js]: I think `injections` is better as a key since it's a noun

File: src/core/instance/inject.js [hunk 2]
```diff
@@ -12,11 +13,16 @@ export function initProvide (vm: Component) {
 }
 
 export function initInjections (vm: Component) {
-  const inject: any = vm.$options.inject
+  const result = resolveInject(vm.$options.inject, vm)
+  extend(vm, result)
+}
+
+export function resolveInject (inject: any, vm: Component) {
```
Comment (yyx990803) [inline src/core/instance/inject.js]: Missing return type annotation `?Object`

File: src/core/vdom/create-component.js (hunk 1 above)
Comment (Kingwl) [inline src/core/vdom/create-component.js]: done

File: src/core/instance/inject.js (hunk 2 above)
Comment (Kingwl) [inline src/core/instance/inject.js]: done

**Pull request merged**